service/service_construction.go

```diff
@@ -588,3 +619,176 @@ func (s ConstructionService) ConstructionSubmit(
 		},
 	}, nil
 }
+
+func (s ConstructionService) CreateOperationDescription(
+	operations []*types.Operation,
+) ([]*parser.OperationDescription, error) {
+	if len(operations) != 2 { //nolint:gomnd
+		return nil, fmt.Errorf("invalid number of operations")
+	}
+
+	firstCurrency := operations[0].Amount.Currency
+	secondCurrency := operations[1].Amount.Currency
+
+	if firstCurrency == nil || secondCurrency == nil {
+		return nil, fmt.Errorf("invalid currency on opeartion")
+	}
+
+	if types.Hash(firstCurrency) != types.Hash(secondCurrency) {
+		return nil, fmt.Errorf("currency info doesn't match between the operations")
+	}
+
+	if types.Hash(firstCurrency) == types.Hash(mapper.AvaxCurrency) {
+		return s.createOperationDescriptionNative(), nil
+	}
+	// Not Native Avax, we require contractInfo in metadata
+	if firstCurrency.Metadata[mapper.ContractAddressMetadata] == nil ||
+		secondCurrency.Metadata[mapper.ContractAddressMetadata] == nil {
+		return nil, fmt.Errorf("non-native currency must have contractAddress in metadata")
+	}
+
+	contractAddress := firstCurrency.Metadata[mapper.ContractAddressMetadata]
```

Comment:
nit: make sure first/second match before using first

Reply:
We check the hash of the currencies to match above, which checks that the metadata matches as well